eis_toolkit/spatial_analyses/cell_based_association.py

```diff
@@ -0,0 +1,511 @@
+                [
+                    (x_left_origin, y_top),
+                    (x_right_origin, y_top),
+                    (x_right_origin, y_bottom),
+                    (x_left_origin, y_bottom),
+                ]
+            )
+            data += [[i + j * cols, i, j, polygon]]
+            y_top = y_top - cell_height
+            y_bottom = y_bottom - cell_height
+        x_left_origin = x_left_origin + cell_width
+        x_right_origin = x_right_origin + cell_width
+    result = pd.DataFrame(data, columns=["cell_id", "x", "y", "geometry"])
+    result.set_index("cell_id", inplace=True)
+
+    return gpd.GeoDataFrame(result, geometry="geometry", crs=working_map.crs)
+
+
+def from_csv(input_csv_file_path: str) -> gpd.GeoDataFrame:
+    """Intermediate utility.
+
+    Read a CBA grid from CSV file. CSV file must contain "geometry"
+    column in 'Well Known Text' (WKT). If a sequence like "__epsg_XXXXX"
+    is contained in the file name, it is used to indicate the coordinate
+    system (CRS).
+
+    Args:
+        input_csv_file_path: CSV filename.
+
+    Returns:
+        Corresponding CBA object.
+    """
+
+    tmp = input_csv_file_path.split("__")
+    crs = tmp[-1].split(".")[0].replace("_", ":")
+    df = pd.read_csv(input_csv_file_path)
+    df["geometry"] = df["geometry"].apply(wkt.loads)
+    # df.insert(0, "geometry", df["geometry"].apply(wkt.loads))
+    # df["geometry"] = df["geom"].apply(wkt.loads)
+    # df = df.drop("geom", axis=1)
+    cba_grid = gpd.GeoDataFrame(df, geometry="geometry")
+    cba_grid = cba_grid.set_crs(crs)
+    cba_grid.index = df["cell_id"]
+    cba_grid = cba_grid.drop("cell_id", axis=1)
+    cba_grid = cba_grid.astype("int32", errors="ignore")
+    return cba_grid
+
+
+def from_vector_file(cba_vector_file_path: str) -> gpd.GeoDataFrame:
+    """Intermediate utility.
+
+    Read a CBA grid from a vector file (ESRI shapefile, geojson...).
+    All attributes (and all values of each attribute) are kept.
+    The geometries contained in the file are assumed to represent a regular
+    square grid.
+
+    Args:
+        cba_vector_file_path: input vector file.
+
+    Returns:
+        Corresponding CBA object.
+    """
+    geodata = gpd.GeoDataFrame.from_file(cba_vector_file_path)
+    cba_grid = gpd.GeoDataFrame(geodata, geometry="geometry")
+    cba_grid.set_index("cell_id", inplace=True)
+    cba_grid.crs = geodata.crs
+    cols = cba_grid.columns.tolist()
+    cols = cols[-1:] + cols[:-1]
+    cba_grid = cba_grid[cols]
+    cba_grid = cba_grid.astype("int32", errors="ignore")
+    return cba_grid
+
+
+def to_csv(cba: gpd.GeoDataFrame, output_path: str) -> None:
+    """Intermediate utility.
+
+    Saves the object in a CSV file.
+
+    Args:
+        output_csv_file_path: name of the saved file suffixed with the
+            CRS (spatial projection system).
+
+    Returns:
+        None
+    """
+
+    crs_txt = f"epsg:{cba.crs.to_epsg()}"
+    tmp = pd.DataFrame(cba)
```

Comment:
Without `copy`, the input `GeoDataFrame` is modified as a side effect of this function.

```suggestion
    tmp = pd.DataFrame(cba.copy())
```